src/TierRegistry.sol

```diff
@@ -0,0 +1,212 @@
+                                 ERRORS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Revert if the caller attempts to rent zero units.
+    error InvalidAmount();
+
+    error InvalidTier();
+
+    error InvalidPrice();
+
+    error InvalidBatchInput();
+
+    error InvalidAddress();
+
+    /// @dev Revert if the caller is not an owner.
+    error NotOwner();
+
+    /*//////////////////////////////////////////////////////////////
+                                 EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    // TODO(aditi): Maybe rename without Purchase in name
+    event PurchasedTier(uint256 indexed fid, uint256 indexed tier, uint256 forDays);
+
+    event RemoveTier(uint256 tier);
+
+    event SetTier(
+        uint256 tier, uint256 minDays, uint256 maxDays, address vault, address paymentToken, uint256 tokenPricePerDay
+    );
+
+    /*//////////////////////////////////////////////////////////////
+                                CONSTANTS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    string public constant VERSION = "2025.05.21";
+    // TODO(aditi): Update the date
+
+    bytes32 internal constant OWNER_ROLE = keccak256("OWNER_ROLE");
+
+    /*//////////////////////////////////////////////////////////////
+                              PARAMETERS
+    //////////////////////////////////////////////////////////////*/
+
+    struct TierInfo {
+        uint256 minDays;
+        uint256 maxDays;
+        address vault;
+        IERC20 paymentToken;
+        // Keep tier a number not enum so we don't need to ugrade contract to add a new one.
+        uint256 tokenPricePerDay;
+        bool isActive;
+    }
+
+    mapping(uint256 => TierInfo) public tierInfoByTier;
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Set the price feed, uptime feed, and initial parameters.
+     *
+     * @param _initialOwner                  Initial owner address.
+     */
+    constructor(
+        address _initialOwner
+    ) {
+        _grantRole(OWNER_ROLE, _initialOwner);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               MODIFIERS
+    //////////////////////////////////////////////////////////////*/
+
+    modifier onlyOwner() {
+        if (!hasRole(OWNER_ROLE, msg.sender)) revert NotOwner();
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        STORAGE RENTAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @inheritdoc ITierRegistry
+     */
+    function purchaseTier(uint256 fid, uint256 tier, uint256 forDays, address payer) external whenNotPaused {
```

Comment:
Sorry, misunderstood this earlier. An arbitrary `payer` here is unsafe. If you've approved the contract for more than the price of a single tier, I can call this with your address as the payer and my fid as the beneficiary, spending your whole allowance on myself. This needs to pull from `msg.sender`.